test/e2e/reconciler/namespace/controller_test.go

```diff
@@ -73,13 +74,148 @@ func TestNamespaceScheduler(t *testing.T) {
+					ObjectMeta: metav1.ObjectMeta{
+						GenerateName: "e2e-nss-1-",
+					},
+					Spec: clusterv1alpha1.ClusterSpec{
+						KubeConfig: string(server1Kubeconfig),
+					},
+				}, metav1.CreateOptions{})
+				require.NoError(t, err, "failed to create cluster1")
+
+				err = server.expect(namespace, scheduledMatcher(cluster1.Name))
+				require.NoError(t, err, "did not see namespace marked scheduled for cluster1 %q", cluster1.Name)
+
+				t.Log("Create a new cluster")
+
+				server2RawConfig, err := f.Servers[physicalCluster2Name].RawConfig()
+				require.NoError(t, err, "failed to get server 2 raw config")
 
+				server2Kubeconfig, err := clientcmd.Write(server2RawConfig)
+				require.NoError(t, err, "failed to get server 2 kubeconfig")
+
+				cluster2, err := server.clusterClient.Create(ctx, &clusterv1alpha1.Cluster{
+					ObjectMeta: metav1.ObjectMeta{
+						GenerateName: "e2e-nss-2-",
+					},
+					Spec: clusterv1alpha1.ClusterSpec{
+						KubeConfig: string(server2Kubeconfig),
+					},
+				}, metav1.CreateOptions{})
+				require.NoError(t, err, "failed to create cluster2")
+
+				t.Log("Delete the old cluster and expect the namespace to end up scheduled to the new cluster.")
+
+				err = server.clusterClient.Delete(ctx, cluster1.Name, metav1.DeleteOptions{})
+				require.NoError(t, err, "failed to delete cluster1 %q", cluster1.Name)
+
+				err = server.expect(namespace, scheduledMatcher(cluster2.Name))
+				require.NoError(t, err, "did not see namespace marked scheduled for cluster2 %q", cluster2.Name)
+
+				t.Log("Delete the remaining cluster and expect the namespace to be marked unscheduled.")
+
+				err = server.clusterClient.Delete(ctx, cluster2.Name, metav1.DeleteOptions{})
+				require.NoError(t, err, "failed to delete cluster2 %q", cluster2.Name)
+
+				err = server.expect(namespace, unschedulableMatcher())
+				require.NoError(t, err, "did not see namespace marked unschedulable")
+			},
+		},
+		{
+			name: "validate namespace with manual scheduling",
+			work: func(ctx context.Context, t *testing.T, server runningServer) {
+				t.Log("Create a cluster and expect the namespace not to be scheduled to it")
+
+				server1RawConfig, err := f.Servers[physicalCluster1Name].RawConfig()
+				require.NoError(t, err, "failed to get server 1 raw config")
+
+				server1Kubeconfig, err := clientcmd.Write(server1RawConfig)
+				require.NoError(t, err, "failed to marshal server 1 kubeconfig")
+
+				cluster, err := server.clusterClient.Create(ctx, &clusterv1alpha1.Cluster{
+					ObjectMeta: metav1.ObjectMeta{
+						GenerateName: "e2e-nss-",
+					},
+					Spec: clusterv1alpha1.ClusterSpec{
+						KubeConfig: string(server1Kubeconfig),
+					},
+				}, metav1.CreateOptions{})
+				require.NoError(t, err, "failed to create cluster1")
+
+				t.Log("Create a namespace with manual scheduling and expect it to be marked unscheduled")
+
+				namespace, err := server.client.CoreV1().Namespaces().Create(ctx, &corev1.Namespace{
```

Comment:
I hesitated on that also :) That new "test case" can definitely be added to the existing one. I went for the "single responsibility" principle (if such thing applies to test cases :)), to avoid going down the road of writing a giant test case, but that may have been a premature decision.

That being said, my understanding is that the biggest contributor to total execution time is the creation and tear down of the servers, and adding an extra test case is negligible.

Reply:
I think that's fine for now. Note that corner-cases will ideally be validated in unit tests as per #559.